# CSCI 4741 - Rainbow Tables
#### Team Members:
    Will Rooney
    Howard Van Dam
    Jonathan Trejo

<b>To-Do: Describe Rainbow Tables and their use</b>

<b>Chain calculation example: To-Do: elaborate on chain creation</b>

In [1]:
import RainbowTable
rainbowTable = RainbowTable.RainbowTable()
print(rainbowTable.getExampleChain(plaintext='1234', k=5))

     1234 --> 81dc9bdb52d04dc20036dbd8313ed055 --> R1
 --> 3055 --> 2836be05e71a8f34902a6e6b37350134 --> R2
 --> 0135 --> a2a68c298479b64a8536d0fd7c479ee1 --> R3
 --> 4793 --> b28d8e3ccb1ff9d46fa2d9e70222c9e7 --> R4
 --> 2300 --> 46a558d97954d0692411c861cf78ef79 --> R5 --> 7883


<b>To-Do: Describe the chain above</b>

<b>Input a password:</b>

In [2]:
password = '1234'

<b>Create the MD5 hash:</b>

In [3]:
import hashlib
passHash = hashlib.md5(password.encode()).hexdigest()

Result:

In [4]:
print passHash

81dc9bdb52d04dc20036dbd8313ed055


Load a pre-computed rainbow table:

In [5]:
rainbowTable.load('rainbowTable_len4_k100.txt')

The Rainbow Table:

In [6]:
for start,end in rainbowTable.table.items(): print start,'\t',end

0239 	7462
0238 	6749
0038 	3849
0133 	4092
1780 	6413
0232 	3080
0037 	5674
0031 	4835
0139 	4799
0033 	7754
0032 	6838
0131 	1385
0543 	1123
1145 	6487
0035 	8063
0142 	0051
0144 	9780
0309 	3719
0146 	0777
0149 	3696
0148 	8572
0305 	1410
0304 	8928
0231 	7094
0203 	4085
0049 	7990
0125 	9625
0122 	1201
0121 	8542
0040 	1278
0138 	3391
0043 	1317
1797 	2453
0045 	4577
0128 	2633
0235 	9270
0430 	7242
0715 	2252
1875 	0408
0030 	6221
0337 	2654
0058 	4750
0116 	0646
0115 	6431
0053 	3546
0052 	6174
0051 	9749
0119 	1547
0057 	7418
0055 	9243
0021 	7122
0026 	5294
1904 	6601
0136 	8674
0326 	7965
0289 	0263
0282 	6620
0328 	9912
0280 	0625
1421 	5567
0207 	4213
0990 	4327
0449 	8792
0751 	2911
0267 	4813
0109 	6166
0105 	2925
0104 	6514
0106 	0332
0268 	0619
0100 	0314
0241 	9343
0389 	6878
0903 	2797
0865 	6953
7717 	2425
1663 	8405
0295 	8740
0513 	5713
0451 	8611
0199 	9606
0740 	7596
0456 	5190
0194 	1994
0276 	5673
0272 	1894
0016 	5934
0170 	8074
0172 	6619
0173 	7551
0175 	7735

Attempt to crack the password:

In [7]:
if rainbowTable.crack(passHash):
    print 'Success! Your password is:',rainbowTable.password
else:
    print 'Attack failed.'

Attack failed.


If the attack failed, let's load another rainbow table that has a similar success rate:

In [8]:
rainbowTable.load('rainbowTable_len4_k99.txt')

if rainbowTable.crack(passHash):
    print 'Success! Your password is:',rainbowTable.password
else:
    print 'Attack failed.'

Success! Your password is: 1234


Full example where the password we are trying to find is '01234':

In [9]:
import timeit
start = timeit.default_timer()

RainbowTable.crackPassExample()

stop = timeit.default_timer()
print 'Execution time:',format(stop-start,'3.3f'),'seconds\n'

Attempting to crack 3 character password with hash: 4100c4d44da9177247e44a5fc1546778
Attack failed: trying new table
Three character password attack failed

Attempting to crack 4 character password with hash: 4100c4d44da9177247e44a5fc1546778
Attack failed: trying new table
Four character password attack failed

Attempting to crack 5 character password with hash: 4100c4d44da9177247e44a5fc1546778
Found password: 01234 

Execution time: 0.504 seconds

